mantidimaging/core/operations/nan_removal/nan_removal.py

```diff
@@ -87,37 +96,20 @@ def register_gui(form: 'QFormLayout', on_change: Callable, view: 'BaseMainWindow
 
         return {"mode_field": mode_field, "replace_value_field": replace_value_field}
 
+    @staticmethod
+    def _nan_to_median(data: np.ndarray, size: int, edgemode: str):
+        nans = np.isnan(data)
+        if np.any(nans):
+            median_data = np.where(nans, -np.inf, data)
+            median_data = median_filter(median_data, size=size, mode=edgemode)
+            data = np.where(nans, median_data, data)
+            if np.any(data == -np.inf):
+                data = np.where(np.logical_and(nans, data == -np.inf), np.nan, data)
+
+        return data
+
     @staticmethod
     def execute_wrapper(mode_field=None, replace_value_field=None):
         mode_value = mode_field.currentText()
         replace_value = replace_value_field.value()
         return partial(NaNRemovalFilter.filter_func, replace_value=replace_value, mode_value=mode_value)
-
-
-def _nan_to_median(data: np.ndarray, size: int, edgemode: str):
-    nans = np.isnan(data)
-    if np.any(nans):
```

Comment:
To be added to code comments
```
We need to convert any NaNs in the array into -inf, to prevent the median filter giving bad results. Individual NaN pixels will be removed, but blocks of NaN pixels may remain as blocks of -inf. So afterwards we need to convert any remaining -inf back to NaN.
```